ros2doctor/ros2doctor/api/package.py

```diff
@@ -0,0 +1,112 @@
+# Copyright 2019 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+from ament_index_python import get_packages_with_prefixes
+from catkin_pkg.package import InvalidPackage
+from catkin_pkg.package import parse_package
+from ros2doctor.api import DoctorCheck
+from ros2doctor.api import DoctorReport
+from ros2doctor.api import Report
+from ros2doctor.api import Result
+
+import rosdistro
+
+
+def get_ros_packages_info() -> dict:
+    """
+    Return all current distro's packages info using rosdistro API.
+
+    :return: a dictionary contains package name, release, source and status
+    """
+    distro_name = os.environ.get('ROS_DISTRO')
+    distro_name = distro_name.lower()
+    url = rosdistro.get_index_url()
+    i = rosdistro.get_index(url)
+    distro_data = rosdistro.get_distribution(i, distro_name).get_data()
+    return distro_data.get('repositories')
+
+
+class PackageCheck(DoctorCheck):
+    """Check local package versions against release versions on rosdistro."""
+
+    def category(self):
+        return 'package'
+
+    def check(self):
+        """Check packages within the directory where command is called."""
+        result = Result()
+        try:
+            distro_packages_info = get_ros_packages_info()
+        except AttributeError:
+            result.add_error('ERROR: Unable to obtain current distro package information \
+                from rosdistro.')
+            return result
+        local_packages_prefixes = get_packages_with_prefixes()
+        if not local_packages_prefixes or not distro_packages_info:
+            result.add_error('ERROR: local packages info or distro packages info is empty.')
+            return result
+        for package_name, package_prefix in local_packages_prefixes.items():
+            file_path = os.path.join(package_prefix, 'share', package_name)
+            if package_name in distro_packages_info:
```

Comment:
It looks like this won't work for repositories that have more than one package. For example, `rcl_action` is in the repository `rcl`.

```python3
>>> distro_packages_info['rcl']
{'doc': {'type': 'git', 'url': 'https://github.com/ros2/rcl.git', 'version': 'dashing'}, 'release': {'url': 'https://github.com/ros2-gbp/rcl-release.git', 'version': '0.7.7-1', 'tags': {'release': 'release/dashing/{package}/{version}'}, 'packages': ['rcl', 'rcl_action', 'rcl_lifecycle', 'rcl_yaml_param_parser']}, 'source': {'type': 'git', 'url': 'https://github.com/ros2/rcl.git', 'version': 'dashing', 'test_pull_requests': True}, 'status': 'developed'}
```
Note the `rcl` repo has a key `packages` which is a list of packages in the repo, which includes `rcl_action`.

There is no repo `rcl_action`.
```python3
>>> distro_packages_info['rcl_action']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'rcl_action'
```